src/curvefit/uncertainty/draws.py

```diff
@@ -0,0 +1,140 @@
+            - `theta (float)`: between 0 and 1, how much scaling of the residuals to do relative to the prediction mean
+
+        ## `get_draws`
+            Returns generated draws
+            - `group (str)`: Group for which the draws are requested. Defaults to None, in which case returns
+                a dictionary of all draws indexed by groups.
+
+        ## Usage
+        In `ModelRunner.run()`:
+        ```python
+            draws = self.draws.create_draws(
+                data=self.data,
+                model_prototype=self.model,
+                solver_prototype=self.solver,
+                residual_model=self.residual_model,
+                evaluation_space = self.predictive_validity.evaluation_space,
+                theta=self.predictive_validity.theta
+            ).get_draws()
+            '''
+        {end_markdown _ResidualModel}
+        """
+
+    def __init__(self, num_draws, prediction_times, exp_smoothing=None, max_last=None):
+
+        self.num_draws = num_draws
+        self.prediction_times = prediction_times
+        self.exp_smoothing = exp_smoothing
+        self.max_last = max_last
+        self._draws = None
+
+        assert type(self.num_draws) == int
+        assert self.num_draws > 0
+
+        assert type(self.prediction_times) == np.ndarray
+
+        if self.exp_smoothing is not None:
+            assert type(self.exp_smoothing) == float
+            if self.max_last is None:
+                raise RuntimeError("Need to pass in how many of the last models to use.")
+            else:
+                assert type(self.max_last) == int
+                assert self.max_last > 0
+        else:
+            if self.max_last is not None:
+                raise RuntimeError("Need to pass in exponential smoothing parameter.")
+
+    # TODO: Implement exponential smoothing (do we need it?)
+    def create_draws(self,
+                     data: Data,
+                     model_prototype,
+                     solver_prototype,
+                     residual_model: _ResidualModel,
+                     evaluation_space,
+                     theta: float):
+
+        self._draws = {group: [] for group in data.groups}
+        for group in data.groups:
+
+            # Initializing model and doing the "best fit"
+            model = model_prototype.clone()
+            solver = solver_prototype.clone()
+            solver.set_model_instance(model)
+            current_group_data = data._get_df(group=group, copy=False)
+            solver.fit(current_group_data)
+
+            # getting predictions for draws times
+            predictions = solver.predict(
+                t=self.prediction_times,
+                predict_fun=evaluation_space
+            )
+
+            max_t = int(np.round(current_group_data[data.col_t].max()))
+            num_obs = int(sum(~current_group_data[data.col_obs].isnull()))
```

Comment:
This particular thing has not been tested yet, but if we have a sum(boolean_array) it should be a valid counter of True elements.

Reply:
yep this seems like it'll work